fedimint-server/src/consensus/engine.rs

```diff
@@ -472,6 +480,79 @@ impl ConsensusEngine {
             .expect("This is the only place where we write to this key");
     }
 
+    /// Creates a backup of the database in the checkpoint directory. The
+    /// frequency of the checkpoints is controlled by
+    /// `FM_DB_CHECKPOINT_FREQ_ENV`. These checkpoints can be used to
+    /// restore the database in case the federation falls out of consensus.
+    fn checkpoint_database(&self, session_index: u64) {
+        // By default, checkpoint the database every 10 sessions
+        let checkpoint_frequency: u64 = env::var(FM_DB_CHECKPOINT_FREQ_ENV)
+            .unwrap_or(10.to_string())
+            .parse()
+            .expect("FM_DB_CHECKPOINT_FREQ var is invalid");
+
+        if session_index % checkpoint_frequency == 0 {
+            let checkpoint_dir = self.data_dir.join("db_checkpoints");
+
+            if !checkpoint_dir.exists() {
+                if let Err(e) = fs::create_dir_all(&checkpoint_dir) {
+                    warn!(target: LOG_CONSENSUS, ?checkpoint_dir, ?e, "Could not create checkpoint directory");
+                    return;
+                }
+            }
+
+            let session_checkpoint_dir = checkpoint_dir.join(format!("{session_index}"));
+
+            match self.db.checkpoint(&session_checkpoint_dir) {
+                Ok(()) => {
+                    info!(target: LOG_CONSENSUS, ?session_checkpoint_dir, ?session_index, "Created db checkpoint");
+                }
+                Err(e) => {
+                    warn!(target: LOG_CONSENSUS, ?session_checkpoint_dir, ?session_index, ?e, "Could not create db checkpoint");
+                }
+            }
+
+            // Check if any old checkpoints need to be cleaned up
+            match Self::delete_old_database_checkpoints(session_index, &checkpoint_dir) {
+                Ok(session_threshold_index) => {
+                    info!(target: LOG_CONSENSUS, ?session_threshold_index, "Deleted all checkpoints before the threshold");
+                }
+                Err(e) => {
+                    warn!(target: LOG_CONSENSUS, ?e, "Could not delete old checkpoints");
+                }
+            }
+        }
+    }
+
+    /// Iterates through the db checkpoint directory and deletes any checkpoint
+    /// that is less than `session_index` - `session_diff`. By default,
+    /// checkpoints that are older than 50 sessions are deleted.
+    fn delete_old_database_checkpoints(
+        session_index: u64,
+        checkpoint_dir: &Path,
+    ) -> anyhow::Result<u64> {
+        // By default, cleanup any checkpoint that is older than 50 sessions
+        let session_diff: u64 = env::var(FM_DB_CHECKPOINT_SESSION_DIFFERENCE_ENV)
+            .unwrap_or(50.to_string())
+            .parse()
+            .expect("FM_DB_CHECKPOINT_SESSION_DIFFERENCE var is invalid");
+        let session_threshold_index = session_index.saturating_sub(session_diff);
+
+        for entry in fs::read_dir(checkpoint_dir)? {
```

Comment:
Instead of iterating every time, we could use some more adaptive strategy.

E.g. we could generate a random number between 0 and session different, and if it exist, delete it. This would prune most checkpoints, leaving out some snapshot that could be used to faster sync. Wrapping the whole in a TimeReporter probably a good idea too.

Reply:
Just realized that we don't need to iterate at all. Since we know the frequency and the current session, we should know the index of the session that needs to be deleted. Can just check if that exists, delete it, and be done.